src/storage/src/hummock/compactor/fast_compactor_runner.rs

```diff
@@ -76,34 +84,108 @@ impl BlockStreamIterator {
     /// The iterator reads at most `max_block_count` from the stream.
     pub fn new(
         sstable: TableHolder,
-        block_stream: BlockDataStream,
         task_progress: Arc<TaskProgress>,
+        sstable_store: SstableStoreRef,
+        sstable_info: SstableInfo,
+        max_io_retry_times: usize,
+        stats_ptr: Arc<AtomicU64>,
     ) -> Self {
         Self {
-            block_stream,
+            block_stream: None,
             next_block_index: 0,
             sstable,
             iter: None,
             task_progress,
+            sstable_store,
+            sstable_info,
+            io_retry_times: 0,
+            max_io_retry_times,
+            stats_ptr,
         }
     }
 
+    async fn create_stream(&mut self) -> HummockResult<()> {
+        // Fast compact only support the single table compaction.(not split sst)
+        // So we don't need to filter the block_metas with table_id and key_range
+        let block_stream = self
+            .sstable_store
+            .get_stream_for_blocks(
+                self.sstable_info.object_id,
+                &self.sstable.meta.block_metas[self.next_block_index..],
+            )
+            .verbose_instrument_await("stream_iter_get_stream")
+            .await?;
+        self.block_stream = Some(block_stream);
+        Ok(())
+    }
+
     /// Wrapper function for `self.block_stream.next()` which allows us to measure the time needed.
-    async fn download_next_block(&mut self) -> HummockResult<Option<(Bytes, Vec<u8>, BlockMeta)>> {
-        let (data, _) = match self.block_stream.next_block_impl().await? {
-            None => return Ok(None),
-            Some(ret) => ret,
-        };
-        let meta = self.sstable.meta.block_metas[self.next_block_index].clone();
-        let filter_block = self
-            .sstable
-            .filter_reader
-            .get_block_raw_filter(self.next_block_index);
-        self.next_block_index += 1;
-        Ok(Some((data, filter_block, meta)))
+    pub(crate) async fn download_next_block(
+        &mut self,
+    ) -> HummockResult<Option<(Bytes, Vec<u8>, BlockMeta)>> {
+        let now = Instant::now();
+        let _time_stat = scopeguard::guard(self.stats_ptr.clone(), |stats_ptr: Arc<AtomicU64>| {
+            let add = (now.elapsed().as_secs_f64() * 1000.0).ceil();
+            stats_ptr.fetch_add(add as u64, atomic::Ordering::Relaxed);
+        });
+        loop {
+            let ret = match &mut self.block_stream {
+                Some(block_stream) => block_stream.next_block_impl().await,
+                None => {
+                    self.create_stream().await?;
+                    continue;
+                }
+            };
+            match ret {
+                Ok(Some((data, _))) => {
+                    let meta = self.sstable.meta.block_metas[self.next_block_index].clone();
+                    let filter_block = if !cfg!(feature = "failpoints") {
```

Comment:
why do we need a special check here? Is it too complicated to fix the test?

Reply:
just a workaround for failpoints test. The `test_utils` generate fake sst without bloom filter, I will try to remove it tmr.